SANSILL process solvent

darominski opened the pull request:
This PR adds `solvent` as another correction that can be subtracted from the corrected 2D sample data. The solvent is processed separately as a sample, in the same way as the water reference. Then, it is subtracted from the sample data after all other corrections are imposed and after the water reference. 

The changes include a new property in `SANSILLReduction`, `SolventInputWorkspace` that if specified, will be deducted from data. `SANSILLAutoProcess` also has a new property, `SolventFiles`, that is used to provide the processed solvent workspaces to `SANSILLReduction`. An additional column for the `SolventFiles` has been also added to `DrILL` interface rundex.

**To test:**

<!-- Instructions for testing. -->
1. Make sure D11 SystemTest data is in your user data search directories.
2. From the workbench, run  the following script:

```
from mantid.simpleapi import *
from tempfile import gettempdir

MaskBTP(Instrument='D11', Tube='0-6,250-256')
RenameWorkspace(InputWorkspace='D11MaskBTP', OutputWorkspace='mask_vertical')
MaskBTP(Instrument='D11', Pixel='0-4,252-256')
Plus(LHSWorkspace='mask_vertical', RHSWorkspace='D11MaskBTP', OutputWorkspace='edge_masks')
# the edges mask can be used as a default mask for all distances and wavelengths

MaskBTP(Instrument='D11', Tube='115-138,', Pixel='115-139')
RenameWorkspace(InputWorkspace='D11MaskBTP', OutputWorkspace='mask_8m_4_6A_center')
MaskBTP(Instrument='D11', Tube='0-14', Pixel='241-256')
Plus(LHSWorkspace='D11MaskBTP', RHSWorkspace='mask_8m_4_6A_center', OutputWorkspace='mask_8m_4_6A')

MaskBTP(Instrument='D11', Tube='110-146', Pixel='110-146')
RenameWorkspace(InputWorkspace='D11MaskBTP', OutputWorkspace='mask_1m_4_6A_center')
MaskBTP(Instrument='D11', Tube='3-14', Pixel='240-256')
Plus(LHSWorkspace='D11MaskBTP', RHSWorkspace='mask_1m_4_6A_center', OutputWorkspace='mask_1m_4_6A_int')
MaskBTP(Instrument='D11', Tube='77-107', Pixel='245-256')
Plus(LHSWorkspace='D11MaskBTP', RHSWorkspace='mask_1m_4_6A_int', OutputWorkspace='mask_1m_4_6A')

beams = '947,1088'
containers = '973,1003'
container_tr = '988'
beam_tr = '1119'
samples = '975,1005'
sample_tr = '990'
solvents = '1106,1091'
solvent_tr = '1121'
thick = 0.1

# this also tests that already loaded workspace can be passed instead of a file
LoadNexusProcessed(Filename='sens-lamp.nxs', OutputWorkspace='sens-lamp')
# first, process the solvent
SANSILLAutoProcess(
            SampleRuns=solvents,
            BeamRuns=beams,
            DefaultMaskFile='edge_masks',
            MaskFiles='mask_8m_4_6A,mask_1m_4_6A',
            SensitivityMaps='sens-lamp',
            SampleTransmissionRuns=solvent_tr,
            ContainerTransmissionRuns=container_tr,
            TransmissionBeamRuns=beam_tr,
            SampleThickness=thick,
            CalculateResolution='MildnerCarpenter',
            OutputWorkspace='solvents',
            BeamRadius='0.05',
            TransmissionBeamRadius=0.05,
            ClearCorrected2DWorkspace=False,
            StitchReferenceIndex=0
)

tmp_dir = gettempdir()
solvent_dir = [os.path.join(tmp_dir, 'solvent_' + str(i) + '.nxs') for i in range(2)]
SaveNexusProcessed('001106_Sample', solvent_dir[0])
SaveNexusProcessed('001091_Sample', solvent_dir[1])

# reduce samples
SANSILLAutoProcess(
            SampleRuns=samples,
            BeamRuns=beams,
            ContainerRuns=containers,
            DefaultMaskFile='edge_masks',
            MaskFiles='mask_8m_4_6A,mask_1m_4_6A',
            SensitivityMaps='sens-lamp',
            SampleTransmissionRuns=sample_tr,
            ContainerTransmissionRuns=container_tr,
            TransmissionBeamRuns=beam_tr,
            SolventFiles=",".join(solvent_dir),
            SampleThickness=thick,
            CalculateResolution='MildnerCarpenter',
            OutputWorkspace='iq_mult_solvent',
            BeamRadius='0.05',
            TransmissionBeamRadius=0.05,
            StitchReferenceIndex=0
)
```
3. Make sure it runs correctly, that the solvent I(Q) and corrected 2D solvent data is created and that it is subtracted correctly from the sample data.


Fixes #30612 
Part of #30343  

Release notes added.

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DannyHindson commented:
I've just started having a look at this PR. When I run the test script I get an error saying that Mantid can't find the run 1105 that is in the solvents variable:
```
solvents = '1105,1090'
```
I don't see this in the D11 system tests directory. Please can you let me know where I can find this file? It looks like I don't have 1090 either. Thanks

darominski commented:
> I've just started having a look at this PR. When I run the test script I get an error saying that Mantid can't find the run 1105 that is in the solvents variable:
> 
> ```
> solvents = '1105,1090'
> ```
> 
> I don't see this in the D11 system tests directory. Please can you let me know where I can find this file? It looks like I don't have 1090 either. Thanks

Yes indeed, somehow the example in the comment was not consistent with the data used for the system test. I corrected it now, and also fixed a conflict in the release notes.

DannyHindson commented on Testing/SystemTests/tests/framework/SANSILLAutoProcessTest.py:
```diff
@@ -267,6 +367,22 @@ def setUp(self):
         config['logging.loggers.root.level'] = 'Warning'
         config.appendDataSearchSubDir('ILL/D11/')
 
+        # prepare mask for instrument edges first:
```
I see you've added a new system test higher up. What's the reason for this CustomStitching system test changing as well?

darominski replied:
Well spotted as well, that seems to be introduced by merging master, where the test for solvent and custom stitching were mashed together. I removed the redundant mask definitions.

DannyHindson commented on docs/source/algorithms/SANSILLReduction-v1.rst:
```diff
@@ -52,6 +52,7 @@ Different input properties can be specified depending on the value of **ProcessA
 |              | * MaskedInputWorkspace        |                        |
 |              | * ReferenceInputWorkspace     |                        |
 |              | * SensitivityInputWorkspace   |                        |
+|              | * SolventInputWOrkspace       |                        |
```
Please can you change to a lower case 'o'

darominski replied:
Thank you for spotting that, it is corrected now.

DannyHindson reviewed (changes requested):
Thanks for updating the test script. That now runs fine. I've added one question and one v minor change to the documentation

DannyHindson reviewed (approved):
Looks good thanks

gvardany commented:
Please don't merge yet, I shall skim through too.

gvardany commented:
Looks good to me too.
